### DISCUSSION TITLE
GJS and GObject introspection

### DISCUSSION THREAD

**Discussion by AimForNaN** [wmww/gtk4-layer-shell#33] [Q&A]
According to the readme, any language that supports introspection files is supported.

I try to import using GJS.

```js
import LayerShell from 'gi://Gtk4LayerShell?version=1.0';
```

But I get the following error:

```
Error: Requiring Gtk4LayerShell, version 1.0: Typelib file for namespace 'Gtk4LayerShell', version '1.0' not found.
```

The Gtk3 one works, but I want to make use of Gtk4. What am I doing wrong?

**Comment by AimForNaN** [comment]
Reconfigured build.

```
meson setup --prefix=/usr --reconfigure build
```

Seems to load now, but complains about a few things.

```
WARNING: Failed to initialize layer surface, GTK4 Layer Shell may have been linked after libwayland.
** Message: 18:33:15.719: Move gtk4-layer-shell before libwayland-client in the linker options.
** Message: 18:33:15.719: You may be able to fix with without recompiling by setting LD_PRELOAD=/path/to/libgtk4-layer-shell.so
** Message: 18:33:15.719: See https://github.com/wmww/gtk4-layer-shell/blob/main/linking.md for more info

GtkWindow is not a layer surface. Make sure you called gtk_layer_init_for_window ()

GtkWindow is not a layer surface. Make sure you called gtk_layer_init_for_window ()
```
Doing the recommended `LD_PRELOAD` from the linking docs works. However, this is highly inconvenient. Is there a way to get it working without having to prepend it?

**Comment by wmww** [comment]
You gotta make it dlopen `libgtk4-layer-shell.so` before importing GTK4. You don't have to *use* that library for anything, just get it loaded into the process. You can see this is done in both the [Python](https://github.com/wmww/gtk4-layer-shell/blob/a689a399d37d174f2e93442e3adc931a9d8dee90/examples/simple-example.py#L6C7-L6C29) and [Lua](https://github.com/wmww/gtk4-layer-shell/blob/a689a399d37d174f2e93442e3adc931a9d8dee90/examples/simple-example.lua#L9) examples. I don't know exactly how to do it in js but you can probably figure it out.

**Reply by AimForNaN** [reply]
There is a node package `ffi` that could probably be used here, but I also don't want to take that approach. I'll be experimenting with systemd's user environment and hyprland's env feature and see if I can't avoid having to deal with this on the code side. In the long run, I may have to wait for Arch to drop Gtk3 support.
